### PR TITLE
feat(rendering): add per-beat rest display position via beat.restDisplayTone, beat.restDisplayOctave

### LINKED ISSUE #2572: Allow control of rest positioning
See https://github.com/CoderLine/alphaTab/discussions/2559

## Current Behavior

As of today alphaTab purely controls the rest positioning itself using standard rules and collision detection. 

## Needed Behavior

Users should be given custom control on which "pitch height" rests should be shown. This should be an opt-in feature of course. 

MusicXML has some feature to control the display step and octave: 
https://www.w3.org/2021/06/musicxml40/musicxml-reference/elements/rest/

For Guitar Pro we'd have to check if it might contain the same information. 

For alphaTex we need something custom (beat effect). 

## Related problem

As seen in the discussion: if the rests are on the main voice, but notes are placed on the secondary voices, the rests are not displaced automatically on the collision detection. We might need to change the collision detection logic to displace rests after all voices have been layouted. 

If a custom position is set this should overrule collision detection.

### PR DESCRIPTION
**Pull request by roscopeeco**
Adds Beat.restDisplayTone (Tone enum) and Beat.restDisplayOctave to allow specifying the vertical position of a rest on the staff on a per-beat basis. The AlphaTex restdisplaypitch tag (e.g. r.4{restdisplaypitch E4}) sets these values. AccidentalHelper.calculateRestDisplaySteps converts the pitch to rendering steps, aligned to match the legacy RestPosition enum spacing (+0.5).

Includes 77 visual tests covering treble, bass, alto and tenor clefs across staff positions, durations, multi-voice scenarios, and variable line counts. 

This PR only includes TypeScript changes 

### Issues

Fixes #

### Checklist
- [ X] I consent that this change becomes part of alphaTab under it's current or any future open source license
- [ ] Changes are implemented
- [ X] New tests were added 

## Further details
- [ ] This is a breaking change
- [ X] This change will require update of the documentation/website

### CODE REVIEW

File: packages/alphatab/src/importer/alphaTex/AlphaTex1LanguageHandler.ts [hunk 1]
```diff
@@ -1616,6 +1617,25 @@ export class AlphaTex1LanguageHandler implements IAlphaTexLanguageImportHandler
             case 'txt':
                 beat.text = (p.arguments!.arguments[0] as AlphaTexTextNode).text;
                 return ApplyNodeResult.Applied;
+            case 'restdisplaypitch': {
+                const pitchText = (p.arguments!.arguments[0] as AlphaTexTextNode).text.toUpperCase();
+                const toneChar = pitchText[0] as keyof typeof Tone;
```
Comment (Danielku15) [inline packages/alphatab/src/importer/alphaTex/AlphaTex1LanguageHandler.ts]: This code cannot be transpiled to C# and Kotlin. 
You can reuse `alphaTab.model.Tuning` to parse such strings

File: packages/alphatab/src/model/Tone.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+ */
+export enum Tone {
+    C = 0,
+    D = 2,
+    E = 4,
+    F = 5,
+    G = 7,
+    A = 9,
+    B = 11
+}
```
Comment (Danielku15) [inline packages/alphatab/src/model/Tone.ts]: The idea with the enum is good, but it is in conflict with how alphaTab handles tones at all other places. I'd prefer consistency (using raw `numnber` properties) until we can introduce such an enum consistently at all places.

File: packages/alphatab/src/rendering/utils/AccidentalHelper.ts [hunk 3]
```diff
@@ -267,6 +268,28 @@ export class AccidentalHelper {
...
+            case Ottavia._8vb:
+                noteValue += 12;
+                break;
+            case Ottavia._15mb:
+                noteValue += 24;
+                break;
+        }
+        
+        const spelling = ModelUtils.resolveSpelling(bar.keySignature, noteValue, NoteAccidentalMode.Default);
+        return AccidentalHelper.calculateNoteSteps(bar.clef, spelling) + 0.5;
```
Comment (Danielku15) [inline packages/alphatab/src/rendering/utils/AccidentalHelper.ts]: `+ 0.5` seems like a workaround indicating to a problem elsewhere. Can we eliminate or move this to the actual place where things where the need for the workaround gets clear?

File: packages/alphatab/test/visualTests/features/RestPosition.test.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    filler: string;
+    positions: [string, string][];
+    primaryFiller: string;
+    secondaryFiller: string;
+    multiVoicePrimary: [string, string];
+    multiVoiceSecondary: [string, string];
+};
+
+// Staff positions from bottom to top for each clef: [pitch, label]
+const clefs: Record<string, ClefData> = {
```
Comment (Danielku15) [inline packages/alphatab/test/visualTests/features/RestPosition.test.ts]: Please write only code which is safe for transpilation. 

https://alphatab.net/docs/contributing#toolchain

**Review by Danielku15 — DISMISSED** [review]

File: packages/alphatab/src/rendering/utils/AccidentalHelper.ts (hunk 3 above)
Comment (roscopeeco) [reply packages/alphatab/src/rendering/utils/AccidentalHelper.ts]: The issue of adding 0.5 is needed because the glyph baseline data contained in bravura_metadata.json varies per glyph. 
For example, the black note head is perfectly centred vertically [0.5, -0.5] , but the rest glyphs all have different off-centre base lines. 

The 0.5 is used to calculate the perfect positioning of the half note rest, so everything is adjusted by 0.5, 
but visually the 0.5 aligns the half note perfectly. 

Do we want all rests, quarter note, eighth note, 16th note, 32nd etc. notes to be aligned perfectly centred visually on the Y axis,either on the line or in the space, or is it just the half note that needs to be visually perfect? 

If all rests need to be visually perfectly centered, then we'd need to include the Bravura metadata coordinates in the
 calculation. Creating a helper function to do this.

We could call this after calling calculateRestDisplaySteps and add the result, the return value, to replace the 0.5 hard coded value.

Comment (roscopeeco) [reply packages/alphatab/src/rendering/utils/AccidentalHelper.ts]: @Danielku15 any thoughts on this? Also, the other three changes have been made.

Comment (Danielku15) [reply packages/alphatab/src/rendering/utils/AccidentalHelper.ts]: > The issue of adding 0.5 is needed because the glyph baseline data contained in bravura_metadata.json varies per glyph.

This is on purpose. SMuFL specifies that the baseline y=0 should be where the staffline is. 

* https://w3c-cg.github.io/smufl/latest/tables/rests.html#implementation-notes
* https://danielku15.github.io/smufl-viewer/?search=rest

Also for noteheads the baseline indicates the staffline. 

> Do we want all rests, quarter note, eighth note, 16th note, 32nd etc. notes to be aligned perfectly centred visually on the Y axis,either on the line or in the space, or is it just the half note that needs to be visually perfect?

if we have a C4 note, and manually position a rest at C4. the rest glyph will be placed on the same staff-line following the common rest placement and offsets as usual. 

`calculateRestDisplaySteps` should not divert from the general placement logic. adding 0.5 seem to be the wrong choice for this method.

What indeed could be is that some legacy/tech-debt areas in alphaTab interfere with the placement. If that's the case I can have a deeper look at the surroundings. In case of doubts you could always craft a MusicXML file and open it in MusicXML to check the placement logic. 

A MusicXML example where we place rests at the same pitch as notes. 

image

```xml
<note default-x="80.21" default-y="-50">
  <pitch>
    <step>C</step>
    <octave>4</octave>
    </pitch>
  <duration>1</duration>
  <voice>1</voice>
  <type>quarter</type>
  <stem>up</stem>
</note>
<note default-x="117.3" default-y="-20">
  <rest>
    <display-step>C</display-step>
    <display-octave>4</display-octave>
  </rest>
  <duration>1</duration>
  <voice>1</voice>
  <type>quarter</type>
</note>
<note default-x="154.4" default-y="10">
  <pitch>
    <step>A</step>
    <octave>5</octave>
    </pitch>
  <duration>1</duration>
  <voice>1</voice>
  <type>quarter</type>
  <stem>down</stem>
</note>
<note default-x="191.49" default-y="-20">
  <rest>
    <display-step>A</display-step>
    <display-octave>5</display-octave>
  </rest>
  <duration>1</duration>
  <voice>1</voice>
  <type>quarter</type>
</note>
```

Comment (roscopeeco) [reply packages/alphatab/src/rendering/utils/AccidentalHelper.ts]: Okay, as I stated previously, the reason for the offset is to handle the half note rest, which visually does not look correct. If that adjustment is not made, it may not be visually perceptible for some of the other rests but for a half note, it does not position correctly with the standard calculation. 

See the test below for half note positioning as is:
packages/alphatab/test-data/visual-tests/rest-position/alto/rest-position-multi-voice-secondary-only-F3-line-1.png

My solution would be to implement a per rest adjustment value to compensate for the Smuffle glypths off-center placement that would be applied in calculateRestDisplaySteps.

If no adjustment is made then the half note and possibly the whole note rest will look incorrect.

I don't think there's anything wrong with your calculations in alphaTab. It's just an artifact of the off center positioning of the glyphs. 

I'll leave it to you to determine what needs to be done,

**Comment by Danielku15** [comment]
Sorry for the long delay on this PR and thanks a lot for all the foundation work. I finalized the remaining bits and double checked the correct display of all rest durations. You can cross check the display in the visual test results added in this PR. 

Let me know if you find some edge cases which do not look as you'd expected.

**Comment by roscopeeco** [comment]
Okay, thanks for completing this PR. I've taken a look at the tests, and the half note positioning still looks odd to me but if you think this is fine, I'm okay with that and happy to defer to your expertise. 

This solution will give enough granular control to avoid ugly multi-voice rest notation rendering.

**Comment by Danielku15** [comment]
I double checked the display of rests across applications like Dorico, MuseScore etc. and the placement is correct like this. What likely makes it look a bit strange is the fact that alphaTab doesn't draw any ledger lines (yet) for half or full note rests which are shifted. I have it on the radar to maybe use the `restWholeLegerLine` / `restHalfLegerLine` glyphs in such cases. https://w3c-cg.github.io/smufl/latest/tables/rests.html?highlight=rests#rests-ue4e0ue4ff
